README.md
# easyaccess 

**easyaccess**: an enhanced command line SQL interpreter client for astronomical surveys.

**Jump to:**

- [Installation](#installation)
- [Usage](#usage)
- [Configuration](#configuration)
- [FAQ](#faq)
       
<img src="data/help.gif" alt="help_screen" style="width: 70%; margin: auto;">

## Description

`easyaccess` is an enhanced command line interpreter and Python package created to facilitate access to astronomical catalogs stored in SQL Databases. It provides a custom interface with custom commands and was specifically designed to access data from the Dark Energy Survey Oracle database, including autocompletion of tables, columns, users and commands, simple ways to upload and download tables using csv, fits and HDF5 formats, iterators, search and description of tables among others. It can easily be extended to another surveys or SQL databases. The package was completely written in Python and support customized addition of commands and functionalities.

For a short tutorial check [here](https://des.ncsa.illinois.edu/desaccess/docs/easyaccess).

## DES DR1/DR2 access quickstart

To access the DES public data releases, you first need an account, which you can register yourself [here](https://des.ncsa.illinois.edu/desaccess). Once you have login credentials for the public DES data server, you can start `easyaccess` with:

```bash
easyaccess -s desdr
```

## Features

- Nice output format (using pandas)
- Very flexible configuration
- Smart tab autocompletion for commands, table names, column names, and file paths
- Write output results to CSV, TAB, FITS, or HDF5 files
- Load tables from CSV, FITS or HDF5 files directly into DB (memory friendly by using number of rows or memory limit)
- Intrinsic DB commands to describe tables, schema, quota, and more
- easyaccess can be imported as module from Python with a complete Python API
- Run commands directly from command line
- Load SQL queries from a file and/or from the editor
- Show the execution plan of a query if needed
- Python functions can be run in a inline query

## FAQ

We have a running list of [FAQ](FAQ.md) which we will constantly update, please check [here](FAQ.md).

## Contributing

Please take a look at our [Code of Conduct](CODE_OF_CONDUCT.md) and our [contribution guide](CONTRIBUTING.md).

## Citation

If you use `easyaccess` in your research, we encourage you to use this reference [https://arxiv.org/abs/1810.02721](https://arxiv.org/abs/1810.02721) or copy/paste this BibTeX:

```bibtex
@ARTICLE{2018arXiv181002721C,
       author = {{Carrasco Kind}, M. and {Drlica-Wagner}, A. and {Koziol}, A.~M.~G. and
        {Petravick}, D.},
        title = "{easyaccess: Enhanced SQL command line interpreter for astronomical surveys}",
      journal = {arXiv e-prints},
     keywords = {Astrophysics - Instrumentation and Methods for Astrophysics},
         year = 2018,
        month = Oct,
          eid = {arXiv:1810.02721},
        pages = {arXiv:1810.02721},
archivePrefix = {arXiv},
       eprint = {1810.02721},
 primaryClass = {astro-ph.IM},
       adsurl = {https://ui.adsabs.harvard.edu/\#abs/2018arXiv181002721C},
      adsnote = {Provided by the SAO/NASA Astrophysics Data System}
}
```

See also: Carrasco Kind et al., (2019). *easyaccess: Enhanced SQL command line interpreter for astronomical surveys*. **Journal of Open Source Software**, 4(33), 1022, https://doi.org/10.21105/joss.01022

## Installation

Installing `easyaccess` can be a little bit tricky given the external libraries required, in particular the Oracle libraries which are free to use. If you are primarily interested in *using* the `easyaccess` client, we recommend running the Docker image as described below.

### Docker

Running `easyaccess` in Docker is easy. Execute the command below to download and run our published image.

```bash
$ docker run -it --rm \
    registry.gitlab.com/des-labs/kubernetes/easyaccess:latest \
    easyaccess -s desdr

Enter username : 
Enter password : 
Connecting to DB ** desdr ** ...
Loading metadata into cache...
     _______      
     \      \      
  // / .    .\    
 // /   .    _\   
// /  .     / // 
\\ \     . / //  
 \\ \_____/ //   
  \\_______//    DARK ENERGY SURVEY
   `-------`     DATA MANAGEMENT

easyaccess x.y.z. The DESDM Database shell.
_________
DESDR ~> SELECT RA, DEC, MAG_AUTO_G, TILENAME FROM DR2_MAIN sample(0.001) FETCH FIRST 5 ROWS ONLY ;

         RA        DEC  MAG_AUTO_G      TILENAME
1  8.236249 -24.021460   24.450422  DES0032-2415
2  8.084798 -25.715401   26.279263  DES0033-2541
3  8.142266 -35.854926   26.509785  DES0032-3540
4  8.197418 -48.274010   25.243387  DES0030-4831
5  8.107404 -26.313876   24.758778  DES0032-2623
```

Alternatively, you may build and run the image yourself using the included Dockerfile with the commands:

```bash
docker build -t des-easyaccess .
docker run -it --rm des-easyaccess easyaccess -s desdr
```

### Source Installation

`easyaccess` is based heavily on the Oracle python client `cx_Oracle`, you can follow the installation instructions from [here](https://cx-oracle.readthedocs.io/en/latest/user_guide/installation.html#quick-start-cx-oracle-installation). For `cx_Oracle` to work, you will need the Oracle Instant Client packages which can be obtained from [here](https://www.oracle.com/technetwork/database/database-technologies/instant-client/overview/index.html).

Make sure you have these libraries installed before proceeding to the installation of easyaccess, you can try by opening a Python interpreter and type:

```python
import cx_Oracle
```

If you have issues, please check the [Troubleshooting page](https://cx-oracle.readthedocs.io/en/latest/user_guide/installation.html#troubleshooting) or our [FAQ page](FAQ.md).

You can clone this repository and install `easyaccess` with:

```bash
python setup.py install
```

#### Requirements

- [Oracle Client](https://www.oracle.com/technetwork/database/database-technologies/instant-client/overview/index.html) > 11g.2 (External library, no python)
  Check [here](https://www.oracle.com/technetwork/database/database-technologies/instant-client/overview/index.html) for instructions on how to install these libraries
- [cx_Oracle](https://cx-oracle.readthedocs.io/en/latest/index.html)
  - Note that cx_Oracle needs libaio on some Linux systems
  - Note that cx_Oracle needs libbz2 on some Linux systems
- See the `setup.py` file for additional requirements.

## Usage

For a short tutorial and documentation see [here](https://des.ncsa.illinois.edu/desaccess/docs/easyaccess). Note that not all the features are available for public database use.

### Interactive interpreter

Assuming that ```easyaccess``` is in your path, you can enter the interactive interpreter by calling ```easyaccess``` without any command line arguments:

```bash
easyaccess
```

### Command line usage

Much of the functionality provided through the interpreter is also available directly from the command line. To see a list of command-line options, use the ```--help``` option

```bash
easyaccess --help
```

### Running SQL commands

Once inside the interpreter run SQL queries by adding a ";" at the end::

```bash
DESDB ~> select ... from ... where ... ;
```

To save the results into a table add ">" after the end of the query (after ";") and namefile at the end of line

```bash
DESDB ~> select ... from ... where ... ; > test.fits
```

The file types supported so far are: .csv, .tab, .fits, and .h5. Any other extension is ignored.

### Load tables

To load a table it needs to be in a csv format with columns names in the first row
the name of the table is taken from filename or with optional argument --tablename

```bash
DESDB ~> load_table <filename> --tablename <mytable> --chunksize <number of rows to read/upload> --memsize <memory in MB to read at a time>
```

The `--chunsize` and `--memsize` are optional arguments to facilitate uploading big files.

### Load SQL queries

To load SQL queries just run:

```bash
DESDB ~> loadsql <filename.sql>
```
or

```bash
DESDB ~> @filename.sql
```

The query format is the same as the interpreter, SQL statement must end with `;` and to write output files the query must be followed by a standard output redirect `command > outfile`

## Configuration

The configuration file is located at ```$HOME/.easyaccess/config.ini``` but everything can be configured from inside easyaccess type:

```bash
DESDB ~> help config
```

to see the meanings of all the options, and:

```bash
DESDB ~> config all show
```

to see the current values, to modify one value, e.g., the prefetch value

```bash
DESDB ~> config prefetch set 50000
```

and to see any particular option (e.g., timeout):

```bash
DESDB ~> config timeout show
```

## Architecture

We have included a simplified UML diagram describing the architecture and dependencies of `easyaccess` which shows only the different methods for a given class and the name of the file hosting a given class. The main class, `easy_or()`, inherits all methods from all different subclasses, making this model flexible and extendable to other surveys or databases. These methods are then converted to command line commands and functions that can be called inside `easyaccess`. Given that there are some DES specific functions, we have moved DES methods into a separate class `DesActions()`.

![`easyaccess` architecture diagram](paper/classes_simple.png)

## Release and publication

### GitHub release

Create a git tag with the semantic version number and make a release in the GitHub repo based on that git tag.

### Release script

Run the `release.sh` shell script to check that all steps have been executed.

### PyPi

*  Make sure version_tag in `./easyaccess/version.py` matches git repo tag

   ```
   version_tag = (x, y, z) 
   ```
* Make sure you have twine installed on system

   ```
   pip install twine
   ```
* In repository directory:

   1. Create tar.gz source file:
   ```
   python3 setup.py sdist
   ```
   2. Create py2-py3 build wheel
   ```
   python3 setup.py bdist_wheel --universal
   ```
   3. Upload build and source files to PyPi
   ```
   twine upload dist/* 
   ```


CHANGES.md
# Changes

## v.1.4.12
#### 2023-APR-10
- Update fitsio dependency to v1.1.8 to address a bug related to FITS file generation.

## v.1.4.11
#### 2022-DEC-06
- Use 'U' instead of 'S' for strings remove code that tried to decode object to utf-8. See pull request: [use unicode type, fix string writing](https://github.com/des-labs/easyaccess/pull/6)

## v.1.4.10
#### 2022-MAY-28
- Correct version inconsistency
- Make requirements consistent between `setup.py` and `meta.yaml`
- Change references to personal email addresses to the DES help request page
 
## v.1.4.8
#### XXXX-XXX-XX
- Fix issues with hanging after a Oracle Error (#PR 173 and issue #130)
 

## v1.4.7
#### 2019-FEB-21
- Update FAQ (PR #171)
- Remove DB names restriction to allow other DBs (PR #170)

## v1.4.6
#### 2019-JAN-14
- No minor changes to the code, just the general documentation
- Adding Contribution guidelines and Coda of Conduct
- Adding Authors
- Adding FAQ
- Updating Readme and Installation instructions
- Reformatting code
- Update paper and references


## v1.4.5
#### 2018-OCT-04
- Restructuring code structure (PR #151)
- Add paper reference
- Minor fixes in unitests
- Update documentation
- Fix conda issues

## v1.4.4
#### 2018-JUN-15
- Improved version to work with public release DB
- Update dtypes to match Y3 tables (PR #149)
- Fix bytes to unicode issue for py3 when writing h5 files
- Fix bug with Oracle types NUMBER with no scale/digits
- Add desdr database and remove oldoper
- Fix HDF5 output files, (see #140)
- Add compression option to configuration, gzip is applied to .csv,.tab,.fits (latter only for 1 DB trip), bzip2 for hdf5
  fits can be opened in append mode when they are compressed, workaround is to increase prefetch

## v1.4.3
#### 2017-AUG-15
- Reduce the number of tries for connection
- Add options to connect to new desoper and remove oldsci, passwords not longer linked (PR #136)
- Allow `describe_table` to work for materialized views (see #135)

## v1.4.2
#### 2017-MAY-23
- Fix a minor migration bug from version 1.4.1
- Add option to reset password after expiration (see #125)
- Add support for new databases (see #126)
- Update tutorial link
- Fix minor bugs and update commands (see #119 #120)
- Fix bugs regarding cx_Oracle (see #117)

## v1.4.0
#### 2017-APR-16
- Refactor and huge clean up, now bin/ folder has the startup script (see PR #113 and #111)
- Fix metadata and cache lookup for autocompletion (see #92, #94, #93)
- Add estimate number of rows on describe_table when available (see #75)
- Toggle color mode without exiting interpreter (see PR #110)
- 80 character loading bar (see PR #103)
- Added more information to help command (see PR #100 and #109)
- Fix config option and boolean values (see Pr #115)
- Justify comments to the left in describe_table i(#105)
- Added `change_db` option switch between dabatases without logging out (see PR #90 , #86)
- Fixes authentication and des service files
- Added api to access descut services directly (see PR #87)
- Fix readline issues related with some OS X (see #88 and #112)


## v1.3.1
#### 2016-MAY-24
- Adds optional purge to drop_table (see PR #74)
- Fix a bug setting prefetch parameters using config set #76
- Improve import functionality for inline queries (see PR #78)
- Add option to upload files in chunk of memory (in addition to the chunk by rows), using --memsize option.
  This calculates an approximate number of rows to upload on each trip, avoiding memory issues. (see PR #79)

## v1.3.0
#### 2016-MAY-09
- Fixes a bug with python2/3 compatibility (see Pull-request #36)
- Removes ; from query for python API (query_* functions) issue #37
- Added execproc to run sql/pl procedures, see help execproc for more information (see PR #48)
- Fix show_index and describe_table commands and queries (no more repeated rows) (see PR #42 and #43)
- Case insensitive options (PR #45)
- Reorganization of the structure of the code, this way is much simpler to develop (PR #49)
- Conda installation
- In-query python functions (beta) write your own function and call it from query#67
- Query and execution information in FITS header #50
- Fix bug when loading sql from file #55
- Dealing better with trailing white space #52
- Fix several minor issues: #37, #40
- configuration options at command line (reset and configuration per session) with --config
- Improved python API , added some extra functions (pandas_to_db) and imprived docs#68
- Load big files in chunks -- very useful to load big files in chunks without running into memory issues #66
- Fix find_tables (to include owners) and describe_table (data_type) commands
- Added unittest suites #65 #70
- Added --no_refresh option in command line to quick start up


## v1.2.0
#### 2015-OCT-01
- Fixes several issues, like: #35, #34, #30, #26 #25 among others
- Does not clear screen after query
- Added a message when password has not been set in the last 200 days  (added extra columns to whoami as well)
- Fix problem with connection after entering a wrong query which used to hangs in some networks
- Bugfixes (show_index, change display setting without restarting, others)
- Added explain option to see the execution plan, i.e. DESDB ~> SELECT * FROM TABLE; < explain
- Added extra configuration option for column width (useful when printing large text columns)
- Added optional user/password from command line (using --user <user> and --password <password>)
- Added size info to mytables command
- Added append_tables option to append data to existing tables
- Minor changes in uploading tables
- **python 3 compatible**
- **change config.ini folder, from .easyacess/ to .easyaccess (typo)**
- move multiprocess import call depending on whether the loading bar is set

## v1.1.0
#### 2015-APR-28
- **Change pyfits to fitsio**
- Added autocommit as configuration option to commit changes or not by default (default = yes)
- Added timeout (1 min) to initial metadata cache loading

## v1.0.8
#### 2015-APR-27

tests/test_load_table.py
#!/usr/bin/env python
"""
Script for testing table upload.
"""
__author__ = "Alex Drlica-Wagner"
import os
from os.path import splitext
from glob import glob
import subprocess as sub

import numpy as np
import pandas as pd
import fitsio
import easyaccess as ea

BASENAME = 'load_table_test'
BASENAME2 = BASENAME + '2'
BASENAME3 = BASENAME + '3'


def create_test_data():
    r = np.linspace(0, 360, 180)
    d = np.linspace(-90, 90, 90)
    ra, dec = np.meshgrid(r, d)
    dtype = [('ra', float), ('dec', float)]
    return np.rec.fromarrays([ra.flat, dec.flat], dtype=dtype)


def create_test_fits(filename=None, data=None):
    if filename is None:
        filename = BASENAME + '.fits'
    if data is None:
        data = create_test_data()
    fitsio.write(filename, data)
    return filename


def create_test_csv(filename=None, data=None):
    if filename is None:
        filename = BASENAME + '.csv'
    if data is None:
        data = create_test_data()
    df = pd.DataFrame(data)
    df.to_csv(filename, index=False, float_format='%.8f', sep=',')
    return filename


def create_test_tab(filename=None, data=None):
    if filename is None:
        filename = BASENAME + '.tab'
    if data is None:
        data = create_test_data()
    df = pd.DataFrame(data)
    df.to_csv(filename, index=False, float_format='%.8f', sep='\t')
    return filename


if __name__ == "__main__":
    import argparse
    description = __doc__
    parser = argparse.ArgumentParser(description=description)
    args = parser.parse_args()

    data = create_test_data()
    nrows = len(data)

    # Create the data files
    fitsfile = create_test_fits(data=data)
    csvfile = create_test_csv(data=data)
    tabfile = create_test_tab(data=data)

    filenames = [fitsfile, csvfile, tabfile]

    # Try loading through the python interface
    # NOTE: This requires a desservice.ini file
    conn = ea.connect()

    query = 'select * from %s' % BASENAME

    # Complains when the table doesn't exist, we could add:
    # if conn.check_table_exists(BASENAME): conn.drop_table(BASENAME)

    for filename in filenames:
        # First try loading through python interface
        print("*** TESTING PYTHON INTERFACE ***")
        conn.drop_table(BASENAME)
        conn.load_table(filename)
        df = conn.query_to_pandas(query)
        assert len(df) == nrows

        # Then try loading with explicit tablename
        print("*** TESTING PYTHON INTERFACE ***")
        conn.drop_table(BASENAME)
        conn.load_table(filename, BASENAME)
        df = conn.query_to_pandas(query)
        assert len(df) == nrows

        # Then try loading through interactive interface
        print("*** TESTING INTERACTIVE INTERFACE ***")
        conn.drop_table(BASENAME)
        cmd = 'load_table %s' % filename
        conn.onecmd(cmd)
        df = conn.query_to_pandas(query)
        assert len(df) == nrows

        # Then try from the command line
        print("*** TESTING COMMAND LINE INTERFACE ***")
        conn.drop_table(BASENAME)
        cmd = 'easyaccess --load_table %s' % filename
        print(cmd)
        sub.check_call(cmd, shell=True)
        df = conn.query_to_pandas(query)
        assert len(df) == nrows

        # Now try downloading a re-uploading
        print("*** TESTING RE-UPLOAD ***")
        filename2 = BASENAME2 + splitext(filename)[-1]
        conn.query_and_save(query, filename2)
        conn.drop_table(BASENAME2)
        conn.load_table(filename2)
        query2 = 'select * from %s' % (BASENAME2)
        df2 = conn.query_to_pandas(query2)
        assert len(df2) == len(data)

    # Now try grabbing from existing table
    nrows = 100000
    query = 'select RA,DEC from Y1A1_COADD_OBJECTS@DESSCI where rownum <= %s;' % nrows
    for ext in ('.fits', '.csv', '.tab'):
        print("*** DOWNLOADING EXISTING TABLE ***")
        basename = BASENAME + '3'
        filename = basename + ext
        conn.query_and_save(query, filename)

        # Test through python interface
        print("*** TESTING PYTHON INTERFACE ***")
        conn.drop_table(basename)
        conn.load_table(filename)
        df = conn.query_to_pandas('select * from %s' % basename)
        assert len(df) == nrows

        # Then try loading through interactive interface
        print("*** TESTING INTERACTIVE INTERFACE ***")
        conn.drop_table(basename)
        cmd = 'load_table %s' % filename
        conn.onecmd(cmd)
        df = conn.query_to_pandas('select * from %s' % basename)
        assert len(df) == nrows

        # Then try from the command line
        print("*** TESTING COMMAND LINE INTERFACE ***")
        conn.drop_table(basename)
        cmd = 'easyaccess --load_table %s' % filename
        print(cmd)
        sub.check_call(cmd, shell=True)
        df = conn.query_to_pandas('select * from %s' % basename)
        assert len(df) == nrows

    # Clean up
    for table in [BASENAME, BASENAME2, BASENAME3]:
        print("*** DROPPING TABLE %s ***" % table)
        conn.drop_table(table)
    filenames = glob(BASENAME + '*.csv') + glob(BASENAME + '*.fits') + glob(BASENAME + '*.tab')
    for filename in filenames:
        print("*** REMOVING FILE %s ***" % filename)
        os.remove(filename)

    print('\n' + "*** TESTS COMPLETED SUCCESSFULLY ***")


easyaccess/eautils/fileio.py
#!/usr/bin/env python
"""
Module for file input/output with pandas, fitsio, ...

Some useful documentation:
fitsio: https://github.com/esheldon/fitsio
numpy:
pandas:

"""
import os
import datetime
import numpy as np
import pandas as pd
import fitsio
import easyaccess.eautils.dtypes as eatypes
import easyaccess.version as version
import sys

try:
    from termcolor import colored
except ImportError:
    def colored(line, color): return line

PANDAS_DEFS = ('comma separated text', 'space separated tex', 'HDF5 format')
PANDAS_EXTS = ('.csv', '.tab', '.h5')

FITS_DEFS = ('FITS format',)
FITS_EXTS = ('.fits',)

GZIP_EXTS = ('.fits', '.csv', '.tab')

FILE_DEFS = PANDAS_DEFS + FITS_DEFS
FILE_EXTS = PANDAS_EXTS + FITS_EXTS


def get_filename(line):
    """
    Return filename after checking it has the right structure (no extra periods)
    """
    line = line.replace(';', '')
    if line == "":
        print('\nMust include table filename!\n')
        return
    if line.find('.') == -1:
        print(colored('\nError in filename\n', "red"))
        return

    filename = "".join(line.split())
    basename = os.path.basename(filename)
    alls = basename.split('.')
    if len(alls) > 2:
        # Oracle tables cannot contain a '.'
        print("\nDo not use extra '.' in filename\n")
        return

    return filename


def get_chunksize(filename, memory=500):
    """
    Get the approximate number of lines ot be read given memory constrains

    Parameters:
    -----------
    filename : File name
    memory   : Memory in MB to compute the approximate number of rows

    Returns:
    --------
    The number of rows need to be read for each chunk of memory
    """
    base, ext = os.path.splitext(filename)
    check_filetype(ext, FILE_EXTS)

    if ext in PANDAS_EXTS:
        if ext == '.csv':
            sepa = ','
        elif ext == '.tab':
            sepa = None
        elif ext == '.h5':
            return IOError('\nReading HDF5 files by chunks is not supported yet\n')
        temp = pd.read_csv(filename, sep=sepa, nrows=100)
        bytes_per_row = temp.memory_usage(index=True).sum() / 100.
        del temp
    elif ext in FITS_EXTS:
        temp = fitsio.FITS(filename)
        temp_data = temp[1][0:100]
        bytes_per_row = temp_data.nbytes / 100.
        temp.close()
        del temp_data

    return int(memory * 1024**2 / bytes_per_row)


def cutquery(query, length):
    """
    Return query in a list of fixed sized character strings
    """
    return [query[0 + i:length + i] for i in range(0, len(query), length)]


def unrecognized_filetype(filename, types=None):
    """
    Return message about unrecognized file type.

    Parameters:
    -----------
    filename : File name (or extension)

    Returns:
    --------
    msg : Unrecognized file message
    """
    if types is None:
        types = FILE_EXTS
    # Try to split the filename
    base, ext = os.path.splitext(filename)
    # Also allow just the file extension
    if ext == '':
        ext = base

    msg = "Unrecognized file type: '%s'\n" % ext
    msg += "Supported filetypes:\n"
    msg += ' ' + ', '.join("'%s'" % t for t in types)
    return msg


def check_filetype(filename, types=None):
    """
    Check file extension against allowed types.

    Parameters:
    -----------
    filename : Name (or extension) of file

    Returns:
    --------
    True : (Or raises IOError)
    """
    if types is None:
        types = FILE_EXTS
    # Try to split the filename
    base, ext = os.path.splitext(filename)
    # Also allow just the file extension
    if ext == '':
        ext = base

    if ext not in types:
        msg = unrecognized_filetype(ext, types)
        raise IOError(msg)
    else:
        return True


def write_file(filename, data, desc, fileindex=1, mode='w', max_mb=1000, query='', comp=False):
    """
    Write a pandas DataFrame to a file. Append to existing file as
    long as smaller than specified size.  Create a new file (and
    increment fileindex) when file grows too large.

    'fileindex' is 1-indexed for backwards compatibility

    Parameters:
    -----------
    filename : Output base filename (incremented by 'fileindex')
    data :     The DataFrame to write to the file
    desc :     The Oracle data descriptor
    fileindex: The index of the file to write.
    mode :     The write-mode: 'w'=write new file, 'a'=append to existing file
    max_mb :   Maximum file size.
    query :    Query used to generate data
    comp  :    Use compresion (gzip)

    Returns:
    fileindex: The (possibly incremented) fileindex.
    """
    base, ext = os.path.splitext(filename)
    check_filetype(filename, FILE_EXTS)
    fileout = filename
    if comp and ext in GZIP_EXTS:
        fileout += '.gz'
    if mode == 'w':
        header = True
    if mode == 'a':
        if (fileindex == 1):
            thisfile = filename
        else:
            thisfile = base + '_%06d' % fileindex + ext
        if comp and ext in GZIP_EXTS:
            thisfile += '.gz'

            # check the size of the current file
        size = float(os.path.getsize(thisfile)) / (2. ** 20)

        if (size > max_mb):
            # it's time to increment the file
            if (fileindex == 1):
                # this is the first one ... it needs to be moved
                lastfile = base + '_%06d' % fileindex + ext
                if comp and ext in GZIP_EXTS:
                    filename += '.gz'
                    lastfile += '.gz'
                os.rename(filename, lastfile)

            # and make a new filename, after incrementing
            fileindex += 1

            thisfile = base + '_%06d' % fileindex + ext
            if comp and ext in GZIP_EXTS:
                thisfile += '.gz'
            fileout = thisfile
            mode = 'w'
            header = True
        else:
            fileout = thisfile
            header = False

    if ext in PANDAS_EXTS:
        write_pandas(fileout, data, fileindex, mode=mode, header=header, query=query, comp=comp)
    if ext in FITS_EXTS:
        write_fitsio(fileout, data, desc, fileindex, mode=mode, query=query, comp=comp)

    return fileindex


def write_pandas(filename, df, fileindex, mode='w', header=True, query='', comp=False):
    """
    Write a pandas DataFrame to a file. Accepted file extension are
    defined by 'PANDAS_EXTS'.

    Parameters:
    -----------
    filename:  Output filename: '.csv','.tab','.h5'
    df :       DataFrame object
    fileindex: Index of this file (modifies filename based on maxfilesize)
    mode :     Write mode: 'w'=write, 'a'=append
    header :   Write header information

    Returns:
    --------
    None
    """
    base, ext = os.path.splitext(filename.replace('.gz',''))
    check_filetype(filename.replace('.gz',''), PANDAS_EXTS)
    if ext == '.csv':
        if comp:
            df.to_csv(filename, index=False, float_format='%.8f', sep=',',
                      mode=mode, header=header,compression='gzip')
        else:
            df.to_csv(filename, index=False, float_format='%.8f', sep=',',
                      mode=mode, header=header, encoding='utf-8')
    if ext == '.tab':
        if comp:
            df.to_csv(filename, index=False, float_format='%.8f', sep=' ',
                      mode=mode, header=header,compression='gzip')
        else:
            df.to_csv(filename, index=False, float_format='%.8f', sep=' ',
                      mode=mode, header=header, encoding='utf-8')
    if ext == '.h5':
        if mode == 'w':
            append = False
        else:
            append = True
        # get current index
        with pd.HDFStore(filename) as storage:
            try:
                nrows = storage.get_storer('data').nrows
            except:
                nrows = 0
        df.index = pd.Series(df.index) + nrows
        if comp:
            df.to_hdf(filename, 'data', mode=mode, format='t', append=append,
                      data_columns=True, complevel=9, complib='bzip2')
        else:
            df.to_hdf(filename, 'data', mode=mode, format='t', append=append,
                      data_columns=True)


def write_fitsio(filename, df, desc, fileindex, mode='w', query='', comp=False):
    """
    Write a pandas DataFrame to a FITS binary table using fitsio.

    It is necessary to convert the pandas.DataFrame to a numpy.array
    before writing, which leads to some hit in performance.

    Parameters:
    -----------
    filename:  Base output FITS filename (over-write if already exists).
    df :       DataFrame object
    desc :     Oracle descriptor object
    fileindex: Index of this file (modifies filename based on maxfilesize)
    mode :     Write mode: 'w'=write, 'a'=append
    query :    Query used to create file
    comp :     Compression 

    Returns:
    --------
    None
    """
    check_filetype(filename.replace('.gz',''), FITS_EXTS)
    # Create the proper recarray dtypes
    dtypes = []
    for d in desc:
        name, otype = d[0:2]
        if otype == eatypes.or_ov:
            # Assume that Oracle OBJECTVARs are 'f8'
            # Could this be better addressed elsewhere?
            dtypes.append((name, 'f8', len(df[name].values[0])))
            print(d, dtypes[-1])
        elif otype == 'updated':
            dtypes.append((name, df[name].dtype.kind + str(df[name].dtype.itemsize)))
        else:
            dtypes.append((name, eatypes.oracle2fitsio(d)))

    # Create numpy array to write
    arr = np.zeros(len(df.index), dtype=dtypes)

    # fill array
    for d in desc:
        name, otype = d[0:2]
        if otype == eatypes.or_ov:
            arr[name] = np.array(df[name].values.tolist())
        else:
            arr[name][:] = df[name].values

    # write or append...
    if mode == 'w':
        # assume that this is smaller than the max size!
        if os.path.exists(filename):
            os.remove(filename)
        fits = fitsio.FITS(filename, mode='rw')
        created = datetime.datetime.now().strftime('%Y-%b-%d %H:%M:%S')
        fits.write(arr)
        fits[1].write_history('Created by easyaccess ' + version.__version__ + ' on ' + created)
        fits[1].write_comment('Query = ' + query)
        fits.close()
        # fitsio.write(filename, arr, clobber=True)
    elif mode == 'a':
        # just append
        fits = fitsio.FITS(filename, mode='rw')
        fits[1].append(arr)
        fits.close()
    else:
        msg = "Illegal write mode!"
        raise Exception(msg)


def read_file(filename):
    """
    Read an input file with pandas or fitsio.

    Unfortunately, the conversion between pandas and numpy is too slow
    to put data into a consistent framework.

    Accepted file extensions are defined by 'FILE_EXTS'.

    Parameters:
    ----------
    filename : Input filename

    Returns:
    --------
    data    : pandas.DataFrame or fitsio.FITS object
    """
    base, ext = os.path.splitext(filename)
    check_filetype(ext, FILE_EXTS)

    if ext in PANDAS_EXTS:
        data = read_pandas(filename)
    elif ext in FITS_EXTS:
        data = read_fitsio(filename)
    else:
        raise IOError()
    return data


def get_dtypes(df):
    """
    Get dtypes from pandas DataFrame or fitsio.FITS
    """
    if df.file_type == 'pandas':
        dtypes = [df[c].dtype if df[c].dtype.kind != 'O'
                  else np.dtype('S' + str(max(df[c].str.len())))
                  for i, c in enumerate(df)]

    if df.file_type == 'fits':
        dtype = df[1].get_rec_dtype(vstorage='fixed')[0]
        dtypes = [dtype[i] for i, d in enumerate(dtype.descr)]
    return dtypes


def read_pandas(filename):
    """
    Read an input file into a pandas DataFrame.  Accepted file
    extension defined by 'PANDAS_EXTS'.

    Parameters:
    ----------
    filename : Input filename

    Returns:
    --------
    df : pandas.DataFrame object
    """
    # ADW: Pandas does a pretty terrible job of automatic typing
    base, ext = os.path.splitext(filename)
    check_filetype(filename, PANDAS_EXTS)

    try:
        if ext in ('.csv', '.tab'):
            if ext == '.csv':
                sepa = ','
            if ext == '.tab':
                sepa = None
            df = pd.read_csv(filename, sep=sepa, iterator=True)
            iterator = True
        elif ext in ('.h5'):
            df = pd.read_hdf(filename, key='data')  # iterator for hdf in padnas 0.18
            iterator = False
    except:
        msg = 'Problem reading %s\n' % filename
        raise IOError(msg)

    # Monkey patch to grab columns and values
    # List comprehension is faster but less readable
    # dtypes = [df[c].dtype if df[c].dtype.kind != 'O'
    #          else np.dtype('S' + str(max(df[c].str.len())))
    #          for i, c in enumerate(df)]

    # df.ea_get_columns = df.columns.values.tolist
    # df.ea_get_values = df.values.tolist
    # df.ea_get_dtypes = lambda: dtypes
    df.file_type = 'pandas'

    return df, iterator


def read_fitsio(filename):
    """Read an input FITS file into a numpy recarray. Accepted file
    extensions defined by 'FITS_EXTS'.

    Parameters:
    ----------
    filename : Input filename

    Returns:
    --------
    fits : fitsio.FITS object
    """
    check_filetype(filename, FITS_EXTS)
    try:
        fits = fitsio.FITS(filename)
    except:
        msg = 'Problem reading %s\n' % filename
        raise IOError(msg)
    # Monkey patch to grab columns and values
    # dtype = fits[1].get_rec_dtype(vstorage='fixed')[0]
    # dtypes = [dtype[i] for i, d in enumerate(dtype.descr)]

    # fits.ea_get_columns = fits[1].get_colnames
    # fits.ea_get_values = fits[1].read().tolist
    # fits.ea_get_dtypes = lambda: dtypes
    fits.file_type = 'fits'

    # ## # Hack to just get a subset of columns
    # ## x1,x2 = 25,37
    # ##
    # ## fits.ea_get_columns = lambda: fits[1].get_colnames()[x1:x2]
    # print fits.ea_get_columns()
    # ## fits.ea_get_values = lambda: fits[1].read(columns=fits.ea_get_columns()).tolist()
    # ## fits.ea_get_dtypes = lambda: dtypes[x1:x2]

    return fits, True


if __name__ == "__main__":
    import argparse

    description = __doc__
    parser = argparse.ArgumentParser(description=description)
    args = parser.parse_args()
